posts/2021-03-25-Rust-1.51.0.md

```diff
@@ -0,0 +1,211 @@
+
+```rust
+impl<T, const LENGTH: usize> Array<T, LENGTH> {
+    fn last(&self) -> Option<&T> {
+        if LENGTH == 0 {
+        // ^^^^^^^^^^^ This is always `true` for `Array<T, 0>`.
+            None
+        } else {
+            Some(&self.list[LENGTH - 1])
+        }
+    }
+}
+```
+
+Const generics adds an important new tool for library designers in creating new, powerful compile-time safe APIs. If you'd like to learn more about const generics you can also check out the ["Const Generics MVP Hits Beta"][const-generics-blog] blog post for more information about the feature and its current restrictions. We can't wait to see what new libraries and APIs you create!
+
+[const-generics-blog]: https://blog.rust-lang.org/2021/02/26/const-generics-mvp-beta.html
+
+### `array::IntoIter` Stabilisation
+As part of const generics stabilising, we're also stabilising a new API that uses it called `array::IntoIter`. `IntoIter` allows you to create a by value iterator over any array. Previously you had to use `[T]`'s (a slice) iterator implementation which returned `&T`, which you then needed to read and copy. Now you can directly iterate over the values without additional copies. *Note:* that `IntoIter::new` may be deprecated once it is possible for arrays to implement `IntoIterator` directly.
+
+```rust
+fn main() {
+  let array = [1, 2, 3, 4, 5];
+  
+  // Previously
+  for item in array.iter().copied() {
+      println!("{}", item);
+  }
+  
+  // Now
+  for item in std::array::IntoIter::new(array) {
+      println!("{}", item);
+  }
+}
+```
+
+### Cargo's New Feature Resolver
+Dependency management is a hard problem, and one of the hardest parts of it is just picking what *version* of a dependency to use when it's depended on by two different packages. This doesn't just include its version number, but also what features are or aren't enabled for the package. Cargo's default behaviour is to merge features for a single package. Let's say you had a dependency called `foo` with features A and B, which was being used by packages `bar` and `baz`, but `bar` depends on `foo+A` and `baz` depends on `foo+B`. Cargo will merge both of those features and compile `foo` as `foo+AB`. This has a benefit that you only have to compile `foo` once, and then it can reused for both `foo` and `bar`.
+
+However, this also comes with a downside. What if `A` and `B` are incompatible? What if your own code is compatible with `foo` but not `foo+A`? A common example of this in the ecosystem is the optional `std` feature included in many `#![no_std]` crates, that allows crates to provide added functionality when `std` is available. Now imagine you want to use the `#![no_std]` version of `foo` in your `#![no_std]` binary, and use the `foo` at build time in your `build.rs`. If your build time dependency depends on `foo+std`, your binary now also depends on `foo+std`, which means it will no longer compile because `std` is not available for your target platform. 
```

Comment:
I think I would drop A and B being incompatible, as likely the support in this version of Cargo won't help there AFAIK; we mostly help when you have the feature X being incompatible with some host or target, I think? (i.e., only when the dependencies are split across profiles, rather than being able to split a dependency in two when compiling for the same target).

Reply:
Yea, maybe the second sentence could be rewritten to:

> What if a feature enabled in a build-dependency is not compatible with the target you are building for?